lib/requestContextPlugin.js

```diff
@@ -10,7 +10,7 @@ function plugin(fastify, opts, next) {
   fastify.decorate('requestContext', requestContext)
   fastify.decorateRequest('requestContext', requestContext)
 
-  fastify.addHook('onRequest', (req, res, done) => {
+  fastify.addHook(opts.hook || 'onRequest', (req, res, done) => {
```

Comment:
I think we might want to run this on multiple hooks as well, maybe it should be an array?

Reply:
Added, but what is the use-case for that? Why would you want to reinitialize storage on multiple steps, if just the earliest one should suffice? Also I'm thinking of switching default to `preValidation`, as it seems to be logical to init it as early as possible. Thoughts?

Reply:
`onRequest` happens _before_ `preValidation`.

I don't think it should be reinitialized, but probably something must be done to retain the context.

Reply:
I digged deeper, and yes, we are definitely losing the context after body-parser is triggered, which actually seems to be a problem that was around a while and was breaking CLS in the past as well. Still investigating what the options are.

Reply:
This looks like a variation of https://github.com/nodejs/node/issues/34401
What are we using for body parsing in fastify? some kind of custom in-house solution? is there some sort of pooling involved there?

Reply:
@vdeturckheim Sorry for bothering you, but have you encountered an issue with ALS in the past where parsing body of an http request was causing async context to be lost?

Reply:
@kibertoad no problem :) I recall something about this but I can't find the PR linked to that. @Flarna @Qard and @puzpuzpuz do you remember anything about this by any chance?

Reply:
I'm not aware of remaining context loss issues related with reused HTTPParser resources. @kibertoad do you have a minimal reproducer?

Reply:
@puzpuz If you don't mind running service locally, https://github.com/rahul2123/fastify-server is the minimum repro. If you would prefer debugging straight into running test, I can add the one that fails to this PR.

Reply:
OK, let me check the repro.

Reply:
@puzpuzpuz I've added test `does not lose request context after body parsing` to this PR that should make it very straightforward to both reproduce and debug this issue. Alternatively you can check that example I've linked before. Note that the most interesting part of behaviour is that it doesn't just lose previous values, it seems to break storage completely, as even values added after the body-parsing has happened cannot be retrieved from the storage (you can see that all the values in the mentioned test are undefined, and also it's the same problem in https://github.com/rahul2123/fastify-server - values set after body parsing are not retrieved correctly)

Reply:
Did some debugging for the failing test. I can see that both of `onRequest` and `preParsing` are executed after the `als.runWith` call within the "proper" async context (`AsyncResource` implicitly created by ALS). However, `preValidation`, `preHandler` and the route handler are executed as a part of [`req.emit('end')`](https://github.com/nodejs/node/blob/master/lib/_stream_readable.js#L1284) call which happens in a different async context, as `req`/`res` may emit events in a different context. That's more or less the problem discussed [here](https://github.com/nodejs/node/issues/33723) and [here](https://github.com/nodejs/node/issues/34430). Please feel free to join those discussions as an ALS user.

To mitigate the issue you could monkey patch EEs, so that they `emit` in a proper async context. See this plugin as an example: https://github.com/puzpuzpuz/cls-rtracer/blob/master/src/rtracer.js#L73

Reply:
@puzpuzpuz Thanks a lot! I'll give it a shot.
@mcollina Do you think that monkey-patching request EE from the plugin is an acceptable solution?

Reply:
> Do you think that monkey-patching request EE from the plugin is an acceptable solution?

Not really.
I think we should attach the context to fastify request and call the `preValidation` callback with that attached.

I _think_ this can be done by manually doing _something_ inside the `'preValidation'` hook. 
@puzpuzpuz do you think it would be possible to create a custom AsyncResource there and restore it in that way?

Reply:
@mcollina @puzpuzpuz Can we attach the whole context, though? Obviously we can instantiate AsyncResource explicitly during `onRequest`, store it in the request and then reuse it from `preValidation` hook, but in order for `runInAsyncScope` to work we need to have a function to call and a receiver for the function, which we do have in case of EventEmitter. What would we use for the same purpose within the `preValidation` hook?

Reply:
I think you can call `done` of the callback form and get that expected result.

Reply:
> I _think_ this can be done by manually doing _something_ inside the `'preValidation'` hook.
> @puzpuzpuz do you think it would be possible to create a custom AsyncResource there and restore it in that way?

If the request object is available in all hooks, then yes, it should be possible to initialize an `AsyncResource` early enough, store it in `req` and then restore it where necessary and, may be, wrap next calls with `asyncResource.runInAsyncScope`. This approach also requires proper integration between ALS and the `AsyncResource`, but that part is trivial.

Reply:
That's possible, the core request is available.

Reply:
@puzpuzpuz @mcollina What am I doing wrong?

```
function plugin(fastify, opts, next) {
  fastify.decorate('requestContext', requestContext)
  fastify.decorateRequest('requestContext', requestContext)
  const hook = opts.hook || 'onRequest'

  if (hook === 'onRequest' || hook === 'preParsing') {
    fastify.addHook('preValidation', (req, res, done) => {
      const asyncResource = req.asyncResource
      asyncResource.runInAsyncScope(done, req)
    })
  }

  fastify.addHook(hook, (req, res, done) => {
    const asyncResource = new AsyncResource('fastify-request-context')
    als.runWith(() => {
      req.asyncResource = asyncResource
      done()
    }, opts.defaultStoreValues)
  })
  next()
}
```

Reply:
@kibertoad you should move `AsyncResource` ctor call inside of the `als.runWith` call. This way it'll be associated with the same store object (I mean the `Map` object created by the `als` wrapper library).

Reply:
@puzpuzpuz Thanks!
@mcollina I believe PR is ready for the final rereview.